File: android-shared-lib/src/test/java/com/skedgo/android/common/model/ModeInfoTest.java
package com.skedgo.android.common.model;

import android.text.TextUtils;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.skedgo.android.common.BuildConfig;

import org.assertj.core.api.Assertions;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricGradleTestRunner;
import org.robolectric.annotation.Config;

import java.util.Comparator;

import static org.assertj.core.api.Assertions.assertThat;

@RunWith(RobolectricGradleTestRunner.class)
@Config(constants = BuildConfig.class, sdk = 21)
public class ModeInfoTest {
  @Test public void parcel() {
    final ModeInfo expected = createModeInfo();
    final ModeInfo actual = ModeInfo.CREATOR.createFromParcel(Utils.parcel(expected));
    Assertions.assertThat(actual)
        .usingComparator(new Comparator<ModeInfo>() {
          @Override
          public int compare(ModeInfo lhs, ModeInfo rhs) {
            return TextUtils.equals(lhs.getAlternativeText(), rhs.getAlternativeText())
                && TextUtils.equals(lhs.getLocalIconName(), rhs.getLocalIconName())
                && TextUtils.equals(lhs.getRemoteIconName(), rhs.getRemoteIconName())
                && TextUtils.equals(lhs.getRemoteDarkIconName(), rhs.getRemoteDarkIconName())
                && TextUtils.equals(lhs.getDescription(), rhs.getDescription())
                && TextUtils.equals(lhs.getId(), rhs.getId())
                ? 0
                : -1;
          }
        })
        .isEqualTo(expected);
    assertThat(actual.getColor()).isEqualTo(expected.getColor());
  }

  @Test public void serializedNames() {
    final ModeInfo info = createModeInfo();
    final JsonObject json = new Gson().toJsonTree(info).getAsJsonObject();
    assertThat(json.has("alt")).isTrue();
    assertThat(json.has("localIcon")).isTrue();
    assertThat(json.has("remoteIcon")).isTrue();
    assertThat(json.has("remoteDarkIcon")).isTrue();
    assertThat(json.has("description")).isTrue();
    assertThat(json.has("identifier")).isTrue();
    assertThat(json.has("color")).isTrue();
  }

  private ModeInfo createModeInfo() {
    final ModeInfo info = new ModeInfo();
    info.setAlternativeText("Taxi");
    info.setLocalIconName("taxi");
    info.setRemoteIconName("taxi");
    info.setRemoteDarkIconName("taxi");
    info.setDescription("Taxi");
    info.setId("ps_tax");
    info.setColor(new ServiceColor(22, 33, 44));
    return info;
  }
}

File: CommonCoreLegacy/src/test/java/com/skedgo/tripkit/routing/ModeInfoTest.kt
package com.skedgo.tripkit.routing

import android.text.TextUtils
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.gson.Gson
import com.skedgo.tripkit.common.model.Utils
import org.assertj.core.api.Java6Assertions
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ModeInfoTest {
    @Test
    fun parcel() {
        val expected = createModeInfo()
        val actual = ModeInfo.CREATOR.createFromParcel(Utils.parcel(expected))
        Java6Assertions.assertThat(actual)
            .usingComparator { lhs, rhs ->
                if (TextUtils.equals(lhs.alternativeText, rhs.alternativeText)
                    && TextUtils.equals(lhs.localIconName, rhs.localIconName)
                    && TextUtils.equals(lhs.remoteIconName, rhs.remoteIconName)
                    && TextUtils.equals(lhs.remoteDarkIconName, rhs.remoteDarkIconName)
                    && TextUtils.equals(lhs.description, rhs.description)
                    && TextUtils.equals(lhs.id, rhs.id)
                ) 0
                else -1
            }
            .isEqualTo(expected)
        Java6Assertions.assertThat(actual.color).isEqualTo(expected.color)
    }

    @Test
    fun serializedNames() {
        val info = createModeInfo()
        val json = Gson().toJsonTree(info).asJsonObject
        Java6Assertions.assertThat(json.has("alt")).isTrue()
        Java6Assertions.assertThat(json.has("localIcon")).isTrue()
        Java6Assertions.assertThat(json.has("remoteIcon")).isTrue()
        Java6Assertions.assertThat(json.has("remoteDarkIcon")).isTrue()
        Java6Assertions.assertThat(json.has("description")).isTrue()
        Java6Assertions.assertThat(json.has("identifier")).isTrue()
        Java6Assertions.assertThat(json.has("color")).isTrue()
    }

    private fun createModeInfo(): ModeInfo {
        val info = ModeInfo()
        info.alternativeText = "Taxi"
        info.localIconName = "taxi"
        info.remoteIconName = "taxi"
        info.remoteDarkIconName = "taxi"
        info.description = "Taxi"
        info.id = "ps_tax"
        info.color = ServiceColor(22, 33, 44)
        return info
    }
}